pkg/kubectl/cmd/taint.go

```diff
@@ -0,0 +1,386 @@
+
+A taint must begin with a letter or number, and may contain letters, numbers, hyphens, dots, and underscores, up to %[1]d characters.`
+	taint_example = `# Update node 'foo' with a taint with key 'dedicated' and value 'special-user' and effect 'NoScheduleNoAdmitNoExecute'.
+# If a taint with that key already exists, its value and effect are replaced as specified.
+kubectl taint nodes foo dedicated=special-user:NoScheduleNoAdmitNoExecute
+# Remove from node 'foo' the taint with key 'dedicated' if one exists.
+kubectl taint nodes foo dedicated-`
+)
+
+func NewCmdTaint(f *cmdutil.Factory, out io.Writer) *cobra.Command {
+	options := &TaintOptions{}
+
+	// retrieve a list of handled resources from printer as valid args
+	validArgs := []string{}
+	p, err := f.Printer(nil, false, false, false, false, false, false, []string{})
+	cmdutil.CheckErr(err)
+	if p != nil {
+		validArgs = p.HandledResources()
+	}
+
+	cmd := &cobra.Command{
+		Use:     "taint NODE NAME KEY_1=VAL_1:TAINT_EFFECT_1 ... KEY_N=VAL_N:TAINT_EFFECT_N",
+		Short:   "Update the taints on a or more nodes",
+		Long:    fmt.Sprintf(taint_long, validation.LabelValueMaxLength),
+		Example: taint_example,
+		Run: func(cmd *cobra.Command, args []string) {
+			if err := options.Complete(f, out, cmd, args); err != nil {
+				cmdutil.CheckErr(err)
+			}
+			if err := options.Validate(args); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(cmd, err.Error()))
+			}
+			if err := options.RunTaint(); err != nil {
+				cmdutil.CheckErr(err)
+			}
+		},
+		ValidArgs: validArgs,
+	}
+	cmdutil.AddValidateFlags(cmd)
+
+	cmdutil.AddPrinterFlags(cmd)
+	cmdutil.AddInclude3rdPartyFlags(cmd)
+	cmd.Flags().StringVarP(&options.selector, "selector", "l", "", "Selector (label query) to filter on")
+	cmd.Flags().BoolVar(&options.overwrite, "overwrite", false, "If true, allow taints to be overwritten, otherwise reject taint updates that overwrite existing taints.")
+	cmd.Flags().BoolVar(&options.all, "all", false, "select all resources in the namespace of the specified resource types")
+	return cmd
+}
+
+func deleteTaintByKey(taints []api.Taint, key string) ([]api.Taint, error) {
+	newTaints := []api.Taint{}
+	found := false
+	for _, taint := range taints {
+		if taint.Key == key {
+			found = true
+			continue
+		}
+		newTaints = append(newTaints, taint)
+	}
+
+	if !found {
+		return nil, fmt.Errorf("taint key=\"%s\" not found.", key)
+	}
+	return newTaints, nil
+}
+
+func reorganizeTaints(accessor meta.Object, overwrite bool, taintsToAdd []api.Taint, removeKeys []string) ([]api.Taint, error) {
+	newTaints := append([]api.Taint{}, taintsToAdd...)
+
+	var oldTaints []api.Taint
+	var err error
+	annotations := accessor.GetAnnotations()
+	if annotations != nil {
+		if oldTaints, err = api.GetTaintsFromNodeAnnotations(annotations); err != nil {
+			return nil, err
+		}
+	}
+
+	// add taints that already existing but not updated to newTaints
+	for _, oldTaint := range oldTaints {
+		existed := false
```

Comment:
calling this variable "existed" is a bit confusing -- better would be exsitsInNew